pytorch_lightning/plugins/training_type/deepspeed.py

```diff
@@ -0,0 +1,323 @@
+            if os.path.exists(config):
+                with open(config) as f:
+                    config = json.load(f)
+            else:
+                raise MisconfigurationException(
+                    f"You passed in a path to a DeepSpeed config but the path does not exist: {config}"
+                )
+        return config
+
+    def pre_dispatch(self):
+        self.set_world_ranks()
+        self.init_ddp_connection(self.global_rank, self.world_size)
+
+        self.init_deepspeed()
+
+        # set warning rank
+        rank_zero_only.rank = self.global_rank
+
+        # set the ranks and devices
+        self.dist.rank = self.global_rank
+        self.dist.device = self.root_device
+        self.barrier()
+
+    def init_deepspeed(self):
+        if not self._config_initialized:
+            self._format_config()
+            self._config_initialized = True
+
+        precision = self.lightning_module.trainer.accelerator_backend.precision
+        model = LightningDeepSpeedModule(pl_module=self.model, precision=precision)
+
+        if self.lightning_module.trainer.training:
+            self._initialize_deepspeed_train(model)
+        else:
+            self._initialize_deepspeed_inference(model)
+
+    def _init_scheduler_optimizer(self):
+        optimizers, schedulers, optimizer_frequencies = self.lightning_module.trainer.init_optimizers(
+            self.lightning_module
+        )
+        if (len(optimizers) != 1) or len(schedulers) > 1:
+            raise MisconfigurationException(
+                "DeepSpeed currently only supports single optimizer, single optional scheduler."
+            )
+        scheduler = schedulers[0]['scheduler'] if len(schedulers) == 1 else None
+        optimizer = optimizers[0]
+        return optimizer, scheduler, optimizer_frequencies
+
+    def _initialize_deepspeed_train(self, model):
+        optimizer, lightning_scheduler, optimizer_frequencies = None, None, None
+        if "optimizer" not in self.config:
+            rank_zero_info(
+                "You have not specified an optimizer or scheduler within the DeepSpeed config."
+                "Using `configure_optimizers` to define optimizer and scheduler."
+            )
+            optimizer, lightning_scheduler, optimizer_frequencies = self._init_scheduler_optimizer()
+        model_parameters = filter(lambda p: p.requires_grad, self.model.parameters())
+        model, optimizer, _, lr_scheduler = deepspeed.initialize(
+            args=SimpleNamespace(local_rank=self.local_rank),
+            model=model,
+            model_parameters=model_parameters,
+            optimizer=optimizer,
+            lr_scheduler=lightning_scheduler,
+            config_params=self.config,
+        )
+
+        # set optimizer for save/load, but deepspeed manages the specific optimizer logic
+        trainer = self.lightning_module.trainer
+        trainer.optimizers = [optimizer]
+        self.model = model
```

Comment:
what is the diff between `self.model` and `self._model` bellow?

Reply:
Good point, I think this is an artifact that should be fixed in the ddp.py file as well

Reply:
yep, early on in the refactor we didn't have a setter yet, so we referred to _model and this seems to be a leftover :)